drivers/espi/espi_emul.c

```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
 
 #include <zephyr/device.h>
 #include <zephyr/drivers/emul.h>
-#include <zephyr/drivers/espi.h>
-#include <zephyr/drivers/espi_emul.h>
-#include "espi_utils.h"
+#include <zephyr/drivers/espi/espi.h>
+#include <zephyr/drivers/espi/espi_emul.h>
+#include <zephyr/drivers/espi/espi_utils.h>
+
+/** Declare the default state of virtual wires */
+const static struct espi_emul_vw_data vw_state_default[] = {
+	{
+		.sig = ESPI_VWIRE_SIGNAL_OOB_RST_ACK,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_WAKE,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_PME,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_SLV_BOOT_DONE,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_ERR_FATAL,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_ERR_NON_FATAL,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_SLV_BOOT_STS,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_SCI,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_SMI,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_RST_CPU_INIT,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_HOST_RST_ACK,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_SUS_ACK,
+		.level = 0,
+	},
+	{
+		.sig = ESPI_VWIRE_SIGNAL_DNX_ACK,
+		.level = 0,
+	},
+};
+
+#define NUMBER_OF_VWIRES ARRAY_SIZE(vw_state_default)
 
-/** Working data for the controller */
+/** Working data for the eSPI peripheral emulator */
 struct espi_emul_data {
-	/* List of struct espi_emul associated with the device */
-	sys_slist_t emuls;
-	/* eSPI host configuration */
+	/**
+	 * This is a pointer to the eSPI host emulator structure, to which the emulated peripheral
+	 * is connected. This peripheral eSPI emulator doesn't emulate whole bus, instead, it
+	 * emulates the eSPI as seen by the MCU, being peripheral connected to the host controller.
+	 */
+	struct espi_emul *host_emul;
+	/** eSPI host configuration */
 	struct espi_cfg cfg;
 	/** List of eSPI callbacks */
 	sys_slist_t callbacks;
+	/** Interrupts state */
+	bool interrupts_en;
+	/** Virtual wires states */
+	struct espi_emul_vw_data vw_state[NUMBER_OF_VWIRES];
+	/** Bit field of enabled channels on eSPI bus */
+	enum espi_channel channels_enabled;
```

Comment:
This is confusing, below you do `channels_enabled & ch` to check if the channel is enabled, the enum is a bit field, should this be a `uint8_t`? Can more than one be set at a time?

Reply:
Hm, it sure can be a misleading. However, the espi subsystem defines the `struct espi_cfg` where you have this enum, with comment in plural form `Supported channels` and also the values of the enums are bits aligned. So yes, multiple values can be set at a time. I've checked and other drivers for chips are using this type in the same way.
E.g. microchip `if (cfg->channel_caps & ESPI_CHANNEL_PERIPHERAL)`